Enable USB Full speed Device controller on LPC55S36 and LPC55S59

mmahadevan108 opened the pull request:
Add support for the Full Speed device controller on LPC55S36 and LPC55S59

mmahadevan108 committed:
soc: nxp_lpc: Add USBFS support

1. Add support for the USB Full Speed controller
2. Add a Kconfig to specify if a dedicated USB
   RAM is available in the SoC.

Signed-off-by: Mahesh Mahadevan <mahesh.mahadevan@nxp.com>

mmahadevan108 committed:
drivers: nxp_usb: Enable USB_DC_HAS_HS_SUPPORT conditionally

Conditionally enable the USB_DC_HAS_HS_SUPPORT as the
USB_DC_NXP_LPCIP3511 controller can have both HS and FS only.

Signed-off-by: Mahesh Mahadevan <mahesh.mahadevan@nxp.com>

mmahadevan108 committed:
dts: lpc55S6x: Add USB FS support

Add support for USB Full speed controller.

Signed-off-by: Mahesh Mahadevan <mahesh.mahadevan@nxp.com>

mmahadevan108 committed:
dts: lpc55S3x: Delete uuid region

uuid region is not present on LPC55S36

Signed-off-by: Mahesh Mahadevan <mahesh.mahadevan@nxp.com>

mmahadevan108 committed:
dts: lpc55S3x: Add USB Full speed support

Add support for USB Full Speed

Signed-off-by: Mahesh Mahadevan <mahesh.mahadevan@nxp.com>

mmahadevan108 committed:
boards: lpcxpresso55s69: Add USB FS support

Add support for the USB FS controller

Signed-off-by: Mahesh Mahadevan <mahesh.mahadevan@nxp.com>

mmahadevan108 committed:
dts: lpc55S6x: Fix the mapping for USB RAM

Fix the mapping for the USB RAM

Signed-off-by: Mahesh Mahadevan <mahesh.mahadevan@nxp.com>

decsny commented on drivers/usb/device/Kconfig:
```diff
@@ -158,7 +158,7 @@ config USB_DC_NXP_EHCI
 config USB_DC_NXP_LPCIP3511
 	bool "LPC USB Device Controller"
-	select USB_DC_HAS_HS_SUPPORT
+	select USB_DC_HAS_HS_SUPPORT if "$(dt_nodelabel_enabled,usbhs)"
```
is there a way to do this without the need for a specifically labelled node in DT?

mmahadevan108 replied:
I am not aware of a way.

decsny commented on soc/arm/nxp_lpc/lpc55xxx/Kconfig.soc:
```diff
@@ -151,4 +151,12 @@ config LPC55XXX_SRAM_CLOCKS
 	  Disable this Kconfig to leave the ram banks untouched out of reset.
 
+config LPC55XXX_USB_RAM
+	bool "Is USB RAM Present?"
```
I think this kconfig should not have a menu prompt. Also, would this be more appropriate as a devicetree property?

mmahadevan108 replied:
Removed the menu prompt. This is Kconfig as we need this to be available in the CMake file.

decsny commented on boards/arm/lpcxpresso55s36/pinmux.c:
```diff
@@ -0,0 +1,28 @@
+	PINCTRL_DT_DEFINE(DT_NODELABEL(usbfs));
+
+	/* Apply pinctrl state directly, since there is no flexio device driver */
+	err = pinctrl_apply_state(PINCTRL_DT_DEV_CONFIG_GET(DT_NODELABEL(usbfs)),
```
why does this not work from DT? there needs to be a flexio driver?

DerekSnell replied:
FlexIO is not related to this, so I think that is an error in the comment

mmahadevan108 replied:
This code has been deleted,

decsny commented on dts/arm/nxp/nxp_lpc55S6x_common.dtsi:
```diff
@@ -314,7 +314,17 @@
 
-	usbhs: usbhs@144000 {
+	usbfs: usbfs@84000 {
+		compatible = "nxp,mcux-usbd";
```
in the commit message here is says "also fix the usb ram mapping" but that seems to happen in a later commit

mmahadevan108 replied:
Thanks for spotting this, I have updated the commit message.

danieldegrasse commented on soc/arm/nxp_lpc/lpc55xxx/soc.c:
```diff
@@ -155,6 +155,41 @@ static ALWAYS_INLINE void clock_init(void)
+	/* enable USB IP clock */
+	CLOCK_EnableUsbfs0DeviceClock(kCLOCK_UsbfsSrcFro, CLOCK_GetFroHfFreq());
+#if defined(FSL_FEATURE_USB_USB_RAM) && (FSL_FEATURE_USB_USB_RAM)
+	for (int i = 0; i < FSL_FEATURE_USB_USB_RAM; i++) {
```
Why not use `memset` here?

mmahadevan108 replied:
Will do.

danieldegrasse commented on boards/arm/lpcxpresso55s36/pinmux.c:
```diff
@@ -0,0 +1,28 @@
+	int err;
+
+	/* Declare pin configuration state for USB pin here */
+	PINCTRL_DT_DEFINE(DT_NODELABEL(usbfs));
```
Is there any reason we could not declare the pin configuration state within the USB DC driver? Even though the driver does not use the device model, it should still be possible to locate this code there and call `pinctrl_apply_state` during init. If no pin control states are specified, the function will of course return an error, but other drivers like the [memc_mcux_flexspi](https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/memc/memc_mcux_flexspi.c#L187) driver simply ignore the return code if it indicates the error was that no pin control state was present.

mmahadevan108 replied:
Thank you, I will add to the USB driver.

imi415 commented:
Hi @mmahadevan108, I have tested this PR on LPCXpresso55S36, it seems the code path with no dedicated USB SRAM has issue which will truncate CONTROL IN data if CONFIGURATION descriptor is above 64 bytes.

I believe this issue should not be related to this PR itself, however AFAIK LPC55(S)36 is the only device with USBFS support without USB SRAM. I believe submitting a issue for this should be more suitable, however since no existing devices will be impacted until this PR is merged, I have to put it here.

I use CDC ACM example, and the length of the CONFIGURATION is 75 bytes.

On a successful CONFIGURATION descriptor request, the transfer is looks like the following:
![usb_success](https://github.com/zephyrproject-rtos/zephyr/assets/10695162/95a8eddb-a80c-43cf-8f38-d18d9ab53f2b)

When the USB SRAM is disabled, the descriptor is truncated and a ZLP is sent:
![usb_fail](https://github.com/zephyrproject-rtos/zephyr/assets/10695162/e64b2cb1-d252-481d-b1e5-dfe270260f74)

On LPC55S69, a simple way to reproduce this is disabling the macro `FSL_FEATURE_USBHSD_USB_RAM_BASE_ADDRESS`, which forces the driver to use allocated pool instead of dedicated SRAM, this way I can reproduce the same issue.

Failed USB enumeration log:
```
[ 5134.914456] usb 1-4.1.2: new full-speed USB device number 65 using xhci_hcd
[ 5135.532379] usb 1-4.1.2: config index 0 descriptor too short (expected 75, got 64)
[ 5135.532392] usb 1-4.1.2: config 1 has an invalid descriptor of length 7, skipping remainder of the config
[ 5135.532400] usb 1-4.1.2: config 1 interface 1 altsetting 0 has 0 endpoint descriptors, different from the interface descriptor's value: 2
[ 5135.916116] usb 1-4.1.2: New USB device found, idVendor=2fe3, idProduct=0001, bcdDevice= 3.03
[ 5135.916132] usb 1-4.1.2: New USB device strings: Mfr=1, Product=2, SerialNumber=3
[ 5135.916138] usb 1-4.1.2: Product: Zephyr CDC ACM sample
[ 5135.916142] usb 1-4.1.2: Manufacturer: ZEPHYR
[ 5135.916146] usb 1-4.1.2: SerialNumber: 0123456789ABCDEF
[ 5136.013519] cdc_acm: probe of 1-4.1.2:1.0 failed with error -22
```

mmahadevan108 committed:
drivers: usb_dc_mcux: Add Pinctrl defines

Add Pinctrl code

Signed-off-by: Mahesh Mahadevan <mahesh.mahadevan@nxp.com>

mmahadevan108 committed:
drivers: usb_dc_mcux: Use CONFIG_USB_DC_NXP_LPCIP3511 define

Use CONFIG_USB_DC_NXP_LPCIP3511 instead of FSL_FEATURE_USB_USB_RAM
as the behaviour variation is based on the SDK driver used and not
if the SoC has a dedicated USB SRAM.

Signed-off-by: Mahesh Mahadevan <mahesh.mahadevan@nxp.com>

mmahadevan108 commented:
> Hi @mmahadevan108, I have tested this PR on LPCXpresso55S36, it seems the code path with no dedicated USB SRAM has issue which will truncate CONTROL IN data if CONFIGURATION descriptor is above 64 bytes.
> 
> I believe this issue should not be related to this PR itself, however AFAIK LPC55(S)36 is the only device with USBFS support without USB SRAM. I believe submitting a issue for this should be more suitable, however since no existing devices will be impacted until this PR is merged, I have to put it here.
> 
> I use CDC ACM example, and the length of the CONFIGURATION is 75 bytes.
> 
> On a successful CONFIGURATION descriptor request, the transfer is looks like the following: ![usb_success](https://user-images.githubusercontent.com/10695162/240847866-95a8eddb-a80c-43cf-8f38-d18d9ab53f2b.png)
> 
> When the USB SRAM is disabled, the descriptor is truncated and a ZLP is sent: ![usb_fail](https://user-images.githubusercontent.com/10695162/240848559-e64b2cb1-d252-481d-b1e5-dfe270260f74.png)
> 
> On LPC55S69, a simple way to reproduce this is disabling the macro `FSL_FEATURE_USBHSD_USB_RAM_BASE_ADDRESS`, which forces the driver to use allocated pool instead of dedicated SRAM, this way I can reproduce the same issue.
> 
> Failed USB enumeration log:
> 
> ```
> [ 5134.914456] usb 1-4.1.2: new full-speed USB device number 65 using xhci_hcd
> [ 5135.532379] usb 1-4.1.2: config index 0 descriptor too short (expected 75, got 64)
> [ 5135.532392] usb 1-4.1.2: config 1 has an invalid descriptor of length 7, skipping remainder of the config
> [ 5135.532400] usb 1-4.1.2: config 1 interface 1 altsetting 0 has 0 endpoint descriptors, different from the interface descriptor's value: 2
> [ 5135.916116] usb 1-4.1.2: New USB device found, idVendor=2fe3, idProduct=0001, bcdDevice= 3.03
> [ 5135.916132] usb 1-4.1.2: New USB device strings: Mfr=1, Product=2, SerialNumber=3
> [ 5135.916138] usb 1-4.1.2: Product: Zephyr CDC ACM sample
> [ 5135.916142] usb 1-4.1.2: Manufacturer: ZEPHYR
> [ 5135.916146] usb 1-4.1.2: SerialNumber: 0123456789ABCDEF
> [ 5136.013519] cdc_acm: probe of 1-4.1.2:1.0 failed with error -22
> ```

@imi415 Thank you for testing. I have updated the PR to fix this.

imi415 commented:
> @imi415 Thank you for testing. I have updated the PR to fix this.
Hi @mmahadevan108, I have tested the new fixes, it works now. Thanks!

danieldegrasse commented on drivers/usb/device/usb_dc_mcux.c:
```diff
@@ -14,6 +14,7 @@
 #include <soc.h>
 #include <zephyr/device.h>
 #include <zephyr/kernel.h>
+#include <zephyr/drivers/pinctrl.h>
```
Add `select PINCTRL` to the Kconfig that enables the USB driver, as this is how we insure pin control is enabled in other drivers: https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/spi/Kconfig.mcux_lpspi#L11

danieldegrasse commented on dts/bindings/usb/nxp,mcux-usbd.yaml:
```diff
@@ -6,7 +6,7 @@ description: |
 compatible: "nxp,mcux-usbd"
 
-include: usb-ep.yaml
+include: [usb-ep.yaml, pinctrl-device.yaml]
```
This `pinctrl-device.yaml` should be updated in the commit enabling pinctrl for the USB driver

danieldegrasse commented on boards/arm/lpcxpresso55s69/pinmux.c:
```diff
@@ -11,7 +11,6 @@
 
 static int lpcxpresso_55s69_pinmux_init(void)
 {
-
```
Nit- this line deletion does not appear to be necessary

danieldegrasse reviewed (commented):
A few small comments, but otherwise this PR looks good

mmahadevan108 committed:
boards: lpcxpresso55s36: Add USBFS support

1. Add support for USB Full Speed controller.
2. Enable USB VBUS pin via pinctrl.

Signed-off-by: Mahesh Mahadevan <mahesh.mahadevan@nxp.com>
